### ISSUE TITLE
Promise#finally: also accept callback that returns Promise<void>

### ISSUE THREAD

**Issue by blair** [microsoft/TypeScript#38752]
**TypeScript Version:**  4.0.0-dev.20200523

**Search Terms:** Promise finally typescript-eslint no-misused-promises

**Code**

```ts
export async function closeDbConnection(): Promise<void> {
    return Promise.resolve();
}

Promise
    .resolve()
    .finally(closeDbConnection);
```

**Expected behavior:**

That typescript-eslint doesn't warn on this code.

**Actual behavior:**

```
test.ts
  7:14  error  Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises
```

One fix that seems to work is changing `lib/lib.es2018.promise.d.ts` from

```ts
    finally(onfinally?: (() => void) | undefined | null): Promise<T>
```

to

```ts
    finally(onfinally?: (() => void | PromiseLike<void>) | undefined | null): Promise<T>
```

```shell
$ mkdir -p /tmp/promise-finally
$ cd /tmp/promise-finally
$ yarn init
$ yarn add -D eslint @typescript-eslint/{eslint-plugin,parser}
$ yarn add -D typescript@next

$ cat >.eslintrc.js <<EOF
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    tsconfigRootDir: __dirname,
    project: ['./tsconfig.json'],
  },
  plugins: [
    '@typescript-eslint',
  ],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-requiring-type-checking',
  ],
};
EOF

$ cat >tsconfig.json <<EOF
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es2019",
    "lib": ["dom", "es2019"],
    "strict": true,
    "strictBindCallApply": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "experimentalDecorators": true,
    "outDir": "build",
    "incremental": true,
    "sourceMap": true,
    "skipLibCheck": true
  },
  "include": [
    "*.ts"
  ],
  "exclude": [
    "node_modules"
  ]
}
EOF

$cat >test.ts <<EOF
export async function closeDbConnection(): Promise<void> {
    return Promise.resolve();
}

Promise
    .resolve()
    .finally(closeDbConnection);
EOF

$ yarn run eslint . --ext .ts
yarn run v1.22.4
$ /tmp/promise-finally/node_modules/.bin/eslint . --ext .ts
=============

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.3.1 <3.8.0

YOUR TYPESCRIPT VERSION: 4.0.0-dev.20200523

Please only submit bug reports when using the officially supported version.

=============

/tmp/promise-finally/test.ts
  7:14  error  Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command
```

**Comment by RyanCavanaugh** [comment]
We don't own or maintain `typescript-eslint`

**Comment by blair** [comment]
@RyanCavanaugh no, but this proposed change is in TypeScript's https://github.com/microsoft/TypeScript/blob/master/lib/lib.es2018.promise.d.ts

**Comment by SeanMcMillan** [comment]
I just ran into this myself. IMHO, `finally` handles return values the same as `then`, and should be typed as such.

**Comment by ephys** [comment]
I have the same problem. This isn't an issue with `typescript-eslint` but with TypeScript's built in type definitions. Could this issue be reopened @RyanCavanaugh ?

**Comment by bogdanb** [comment]
Please reopen this issue. The problem is not just eslint, the problem is that the type declaration doesn’t match the methods meaning.

I’ve seen the argument that TypeScript itself allows a `() => Promise<void>` (or, indeed, `() => anything`) where a `() => void` callback argument is expected. But that doesn’t mean we shouldn’t declare what kind of return values are meaningful. 

The current typing is just as bad as declaring the callback to `Array.filter` as a `(v, i, a) => void`.

**Comment by dmurvihill** [comment]
Please reconsider this issue. The return type for `onFinally` still violates the [documented properties](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-functions-returning-non-void-assignable-to-function-returning-void) of void callbacks:

> Another way to think of this is that a void-returning callback type says "I'm not going to look at your return value, if one exists".
